src/limitedmap.h

```diff
@@ -26,68 +26,76 @@ class limitedmap
     typedef typename std::multimap<V, iterator>::iterator rmap_iterator;
     size_type nMaxSize;
 
+private:
+    rmap_iterator find_rmap_key_value_pair(const_iterator map_it) {
+        auto [it, it_last] = rmap.equal_range(map_it->second);
+        for (; it != it_last; ++it) {
+            if (it->second == map_it) {
+                return it;
+            }
+        }
+        return rmap.end();
+    }
+
 public:
-    limitedmap(size_type nMaxSizeIn = 0) { nMaxSize = nMaxSizeIn; }
+    LimitedMap(size_type nMaxSizeIn = 0) { nMaxSize = nMaxSizeIn; }
     const_iterator begin() const { return map.begin(); }
     const_iterator end() const { return map.end(); }
     size_type size() const { return map.size(); }
     bool empty() const { return map.empty(); }
     const_iterator find(const key_type& k) const { return map.find(k); }
     size_type count(const key_type& k) const { return map.count(k); }
-    void insert(const value_type& x)
-    {
+
+    void insert(const value_type& x) {
+        if (nMaxSize == 0) {
+            return;
+        }
+
+        if (map.size() == nMaxSize) {
+            const V& cur_min = rmap.begin()->first;
+            if (x.second <  cur_min) {
+                return;
+            }
+            map.erase(rmap.begin()->second);
+            rmap.erase(rmap.begin());
+        }
         std::pair<iterator, bool> ret = map.insert(x);
         if (ret.second) {
-            if (nMaxSize && map.size() == nMaxSize) {
-                map.erase(rmap.begin()->second);
-                rmap.erase(rmap.begin());
-            }
-            rmap.insert(make_pair(x.second, ret.first));
+            // in zend, we expect to insert increasing timestamps most of the times, hence the hint.
+            rmap.insert(rmap.end(), make_pair(x.second, ret.first));
```

Comment:
I see a problem here. The deletion of an element due to exceeded capacity is done **before** the actual insertion takes place (which might not if the key already exists).
Sample :
```
    LimitedMap<char,int> lmap(2);
    lmap.insert(std::pair<char,int>('a', 7));
    lmap.insert(std::pair<char,int>('b', 5));
    lmap.insert(std::pair<char,int>('a', 6));
    assert(lmap.find('b') != lmap.end());
```
The key 'b' is gone but we don't have exceeded capacity cause the new 'a' has not been inserted cause it already exist. size() now returns 1 but it should return 2

Also I would change the return type of `insert()` from void to bool to signal true -> the insertion has taken place; false -> no insertion performed as they key already exists

Also don't understand the comment
`// in zend, we expect to insert increasing timestamps most of the times, hence the hint.`

Apparently has nothing to do with `ret.second` nor with increasing values.

Reply:
That's a great catch, thanks! I've fixed the problem with the last commit. I also modified the signature of `insert()` as suggested, and rephrased the comment. I hope it's more clear now.